iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/extensions/SlackTerminalImpl.java

```diff
@@ -56,27 +88,35 @@ public void extendVariantArraySize(int initVariantArraySize, int number, int sou
         terminals.ensureCapacity(terminals.size() + number);
         Terminal sourceTerminal = terminals.get(sourceIndex);
         for (int i = 0; i < number; ++i) {
-            terminals.add(sourceTerminal);
+            addTerminalAndUpdateReferences(sourceTerminal);
         }
     }
 
     @Override
     public void reduceVariantArraySize(int number) {
         for (int i = 0; i < number; i++) {
-            terminals.remove(terminals.size() - 1); // remove elements from the top to avoid moves inside the array
+            removeTerminalAndUpdateReferences(terminals.size() - 1); // remove elements from the top to avoid moves inside the array
         }
     }
 
     @Override
     public void deleteVariantArrayElement(int index) {
-        terminals.set(index, null);
+        setTerminalAndUpdateReferences(index, null);
     }
 
     @Override
     public void allocateVariantArrayElement(int[] indexes, int sourceIndex) {
         Terminal terminalSource = terminals.get(sourceIndex);
         for (int index : indexes) {
-            terminals.set(index, terminalSource);
+            setTerminalAndUpdateReferences(index, terminalSource);
+        }
+    }
+
+    @Override
+    public void onReferencedRemoval(Terminal terminal) {
+        int variantIndex = terminals.indexOf(terminal);
+        if (variantIndex != -1) {
+            terminals.set(variantIndex, null);
         }
     }
```

Comment:
There's the cleanup implementation missing, and in ReferenceTerminalsImpl also

Reply:
Done